<a href="https://colab.research.google.com/github/rosspalmer22/BTECH_adv_python/blob/main/Module_6_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Wide World Importers Invoices Exploratory Data Analysis

Ross Palmer

Wide World Importers sells a variety of different products such as slippers, coffee mugs, and flash drives. The company has extensive records of invoices that have been created during the past several years and has asked that I, the data analyst, to explore the data to determine if there is anything noteworthy that executives should incorporate into their business strategy.

## About the data
The data has been input over the years using several different programs, there are some holes in the data set that need "cleaning". The data will eventually be run through a machine learning model and thus cannot contain any null values.

In [ ]:
import pandas as pd
df = pd.read_csv('invoices.csv')
df

<ipython-input-86-8a490e1e8e4a>:2: DtypeWarning: Columns (12) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('invoices.csv')


InvoiceLineID  InvoiceID  StockItemID  \
0                   1          1         67.0   
1                   2          2         50.0   
2                   3          2         10.0   
3                   4          3        114.0   
4                   5          4        206.0   
...               ...        ...          ...   
230543          36558      11285        139.0   
230544          21110       6502          9.0   
230545          34087      10504        179.0   
230546         103435      31836         20.0   
230547         192963      59503         70.0   

                                              Description  PackageTypeID  \
0                Ride on toy sedan car (Black) 1/12 scale              7   
1       Developer joke mug - old C developers never di...              7   
2                    USB food flash drive - chocolate bar              7   
3                      Superhero action jacket (Blue) XXL              7   
4              Permanent marker black 5mm nib (Black) 5mm              7   
...                                                   ...            ...   
230543                        Furry animal socks (Pink) M             10   
230544                      USB food flash drive - banana              7   
230545              Shipping carton (Brown) 229x229x229mm              7   
230546       DBA joke mug - you might be a DBA if (White)              7   
230547           Ride on toy sedan car (Green) 1/12 scale              7   

        Quantity  UnitPrice  TaxRate  TaxAmount  LineProfit  ExtendedPrice  \
0             10     230.00       15     345.00       850.0        2645.00   
1              9      13.00       15        NaN        76.5        -999.00   
2              9      32.00       15      43.20       180.0         331.20   
3              3      30.00       15        NaN        24.0        -999.00   
4             96       2.70       15        NaN        96.0            NaN   
...          ...        ...      ...        ...         ...            ...   
230543        36       5.00       15      27.00       126.0         207.00   
230544         2      32.00       15        NaN        40.0            NaN   
230545       125       1.05       15        NaN        62.5            NaN   
230546         1      13.00       15       1.95         8.5          14.95   
230547         5     230.00       15        NaN       425.0            NaN   

        LastEditedBy  LastEditedWhen          Customer  
0                  7             NaN     Stuff by Stew  
1                  7  1/1/2013 12:00   Wholesaler Plus  
2                  7             NaN   Wholesaler Plus  
3                  7             NaN   Big Buys Retail  
4                  7  1/1/2013 12:00  Terry's Trinkets  
...              ...             ...               ...  
230543            20             NaN   Wholesaler Plus  
230544            10             NaN     stuff by stew  
230545            16             NaN   Terry's Trinket  
230546            20             NaN   Big Buys Retail  
230547             4             NaN     Stuff by Stew  

[230548 rows x 14 columns]

## Cleaning and Analysis
###The cleaning process will include the following:

- Fix standardization errors, where applicable

- Fix business logic errors, where applicable

- Impute null values, where possible

- Drop columns with many null values, where applicable

- Drop rows with null values, where applicable

- Drop duplicate rows, where applicable

- Identify outliers and fix them as instructed


###I will use the data to answer the following questions, after cleaning has been performed:

- Which customer has the most invoices?

- Which item has the highest total quantity sold?

- Which item has generated the most total profit?

- Which item do customers purchase the most quantity of, on average?

###The information below shows that this dataset has 230547 rows.
###There are 14 columns and 5 of those have null values.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230548 entries, 0 to 230547
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   InvoiceLineID   230548 non-null  int64  
 1   InvoiceID       230548 non-null  int64  
 2   StockItemID     230087 non-null  float64
 3   Description     229876 non-null  object 
 4   PackageTypeID   230548 non-null  int64  
 5   Quantity        230548 non-null  int64  
 6   UnitPrice       230548 non-null  float64
 7   TaxRate         230548 non-null  int64  
 8   TaxAmount       109623 non-null  float64
 9   LineProfit      230548 non-null  float64
 10  ExtendedPrice   113081 non-null  float64
 11  LastEditedBy    230548 non-null  int64  
 12  LastEditedWhen  6 non-null       object 
 13  Customer        230548 non-null  object 
dtypes: float64(5), int64(6), object(3)
memory usage: 24.6+ MB


### The table below shows statistical information about the dataset. This can help identify logical errors such as negative numbers in the TaxRate and ExtendedPrice columns.

In [ ]:
df.describe()

InvoiceLineID      InvoiceID    StockItemID  PackageTypeID  \
count  230548.000000  230548.000000  230087.000000  230548.000000   
mean   114127.066880   35177.386891     110.181827       7.073638   
std     65900.142929   20338.920057      63.722514       0.644528   
min         1.000000       1.000000       1.000000       1.000000   
25%     57062.750000   17570.000000      54.000000       7.000000   
50%    114130.500000   35151.000000     111.000000       7.000000   
75%    171203.250000   52766.000000     165.000000       7.000000   
max    228265.000000   70510.000000     227.000000      10.000000   

            Quantity      UnitPrice        TaxRate      TaxAmount  \
count  230548.000000  230548.000000  230548.000000  109623.000000   
mean       39.252524      45.539361      14.463409     113.328006   
std        55.193934     139.510031       3.904056     221.866696   
min         1.000000       0.660000     -15.000000       0.380000   
25%         5.000000      13.000000      15.000000      14.400000   
50%        10.000000      18.000000      15.000000      34.500000   
75%        60.000000      32.000000      15.000000     129.600000   
max      3150.000000    1899.000000      15.000000   14994.000000   

         LineProfit  ExtendedPrice   LastEditedBy  
count  230548.00000  113081.000000  230548.000000  
mean      375.38894     740.092626      10.800480  
std       753.17041    1705.229089       5.509623  
min      -645.00000    -999.000000       2.000000  
25%        51.00000      89.700000       6.000000  
50%       120.00000     248.400000      11.000000  
75%       390.00000     919.080000      16.000000  
max      9200.00000  114954.000000      20.000000

### This table shows that there are some standardization issues in the dataset with the Customer names. It looks like "Terry's Trickets" has the most invoices, but after the standardization issues are resolved you can see farther down that "Stuff by Stew" actually has the most.

In [ ]:
df.value_counts('Customer')

Customer
Terry's Trinkets       85593
Stuff by Stew          62752
Stuf by Stew           31984
Wholesaler Plus        15735
Big Buys Retail        13461
Terry's Trinket         9483
stuff by stew           7068
Dennis Distribution     4472
Name: count, dtype: int64

In [ ]:
df['Customer'] = df['Customer'].replace({
    "Terry's Trinket": "Terry's Trinkets",
    'Stuf by Stew': 'Stuff by Stew',
    'stuff by stew': 'Stuff by Stew',
})

In [ ]:
df.value_counts('Customer')

Customer
Stuff by Stew          101804
Terry's Trinkets        95076
Wholesaler Plus         15735
Big Buys Retail         13461
Dennis Distribution      4472
Name: count, dtype: int64

### The LastEditedWhen column only has data in 6 of the 230548 rows. This data will not be helpful and my recommendation is that it be dropped from the dataset.

In [ ]:
# I would recommed droping the LastEditedWhen column
df.drop(columns=['LastEditedWhen'], inplace = True)

### With the LastEditedWhen column removed, the columns that now has the most null values is the TaxAmount column.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230548 entries, 0 to 230547
Data columns (total 13 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   InvoiceLineID  230548 non-null  int64  
 1   InvoiceID      230548 non-null  int64  
 2   StockItemID    230087 non-null  float64
 3   Description    229876 non-null  object 
 4   PackageTypeID  230548 non-null  int64  
 5   Quantity       230548 non-null  int64  
 6   UnitPrice      230548 non-null  float64
 7   TaxRate        230548 non-null  int64  
 8   TaxAmount      109623 non-null  float64
 9   LineProfit     230548 non-null  float64
 10  ExtendedPrice  113081 non-null  float64
 11  LastEditedBy   230548 non-null  int64  
 12  Customer       230548 non-null  object 
dtypes: float64(5), int64(6), object(2)
memory usage: 22.9+ MB


### I observed that StockItemID is a single number, and each number has a corresponding Description. Thus, by knowing the StockItemID of a row, its Description can be imputed. However, if both the StockItemID and Description are missing, the item is unknown.

### This code shows that there are 461 rows that have both a null StockItemID and a null Description?

In [ ]:
null_rows = df[df[['StockItemID', 'Description']].isnull().all(axis=1)]

In [ ]:
null_rows.shape[0]

461

### Since there are only 461 rows and it is not practical to impute the data my recommendation is to drop the rows with null values in the Description and StockItemID columns

In [ ]:
df.dropna(subset=['StockItemID', 'Description'], how='all', inplace = True)

### After removing those rows we are left with 230087 rows.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 230087 entries, 0 to 230547
Data columns (total 13 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   InvoiceLineID  230087 non-null  int64  
 1   InvoiceID      230087 non-null  int64  
 2   StockItemID    230087 non-null  float64
 3   Description    229876 non-null  object 
 4   PackageTypeID  230087 non-null  int64  
 5   Quantity       230087 non-null  int64  
 6   UnitPrice      230087 non-null  float64
 7   TaxRate        230087 non-null  int64  
 8   TaxAmount      109408 non-null  float64
 9   LineProfit     230087 non-null  float64
 10  ExtendedPrice  112858 non-null  float64
 11  LastEditedBy   230087 non-null  int64  
 12  Customer       230087 non-null  object 
dtypes: float64(5), int64(6), object(2)
memory usage: 24.6+ MB


### I noticed there are negative values in the TaxRate column.

In [ ]:
df.describe()

InvoiceLineID      InvoiceID    StockItemID  PackageTypeID  \
count  230087.000000  230087.000000  230087.000000  230087.000000   
mean   114131.169266   35178.644621     110.181827       7.073468   
std     65894.369187   20337.134754      63.722514       0.644395   
min         1.000000       1.000000       1.000000       1.000000   
25%     57077.500000   17574.000000      54.000000       7.000000   
50%    114131.000000   35151.000000     111.000000       7.000000   
75%    171196.500000   52764.000000     165.000000       7.000000   
max    228265.000000   70510.000000     227.000000      10.000000   

            Quantity      UnitPrice        TaxRate      TaxAmount  \
count  230087.000000  230087.000000  230087.000000  109408.000000   
mean       39.247267      45.532654      14.463268     113.294589   
std        55.193231     139.461436       3.904501     221.796355   
min         1.000000       0.660000     -15.000000       0.380000   
25%         5.000000      13.000000      15.000000      14.400000   
50%        10.000000      18.000000      15.000000      34.500000   
75%        60.000000      32.000000      15.000000     129.600000   
max      3150.000000    1899.000000      15.000000   14994.000000   

          LineProfit  ExtendedPrice   LastEditedBy  
count  230087.000000  112858.000000  230087.000000  
mean      375.314843     739.846178      10.799919  
std       753.072527    1704.618239       5.509919  
min      -645.000000    -999.000000       2.000000  
25%        51.000000      89.700000       6.000000  
50%       120.000000     248.400000      11.000000  
75%       390.000000     919.080000      16.000000  
max      9200.000000  114954.000000      20.000000

In [ ]:
df['TaxRate'].value_counts()

TaxRate
 15    225106
-15      3934
 10      1035
-10        12
Name: count, dtype: int64

In [ ]:
df['TaxRate'].unique()

array([ 15, -15,  10, -10])

### Based on your domain knowledge and knowing that tax rates should not be a negative number my recommendation is to turn the negative tax rates into positive tax rates by multiplying the column by -1.

In [ ]:
df.loc[df['TaxRate'] < 0, 'TaxRate'] *= -1

### Now we can see that the negative values have been removed and average tax rate has gone from 14.46 to 14.97.

In [ ]:
df.describe()

InvoiceLineID      InvoiceID    StockItemID  PackageTypeID  \
count  230087.000000  230087.000000  230087.000000  230087.000000   
mean   114131.169266   35178.644621     110.181827       7.073468   
std     65894.369187   20337.134754      63.722514       0.644395   
min         1.000000       1.000000       1.000000       1.000000   
25%     57077.500000   17574.000000      54.000000       7.000000   
50%    114131.000000   35151.000000     111.000000       7.000000   
75%    171196.500000   52764.000000     165.000000       7.000000   
max    228265.000000   70510.000000     227.000000      10.000000   

            Quantity      UnitPrice        TaxRate      TaxAmount  \
count  230087.000000  230087.000000  230087.000000  109408.000000   
mean       39.247267      45.532654      14.977248     113.294589   
std        55.193231     139.461436       0.336518     221.796355   
min         1.000000       0.660000      10.000000       0.380000   
25%         5.000000      13.000000      15.000000      14.400000   
50%        10.000000      18.000000      15.000000      34.500000   
75%        60.000000      32.000000      15.000000     129.600000   
max      3150.000000    1899.000000      15.000000   14994.000000   

          LineProfit  ExtendedPrice   LastEditedBy  
count  230087.000000  112858.000000  230087.000000  
mean      375.314843     739.846178      10.799919  
std       753.072527    1704.618239       5.509919  
min      -645.000000    -999.000000       2.000000  
25%        51.000000      89.700000       6.000000  
50%       120.000000     248.400000      11.000000  
75%       390.000000     919.080000      16.000000  
max      9200.000000  114954.000000      20.000000

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 230087 entries, 0 to 230547
Data columns (total 13 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   InvoiceLineID  230087 non-null  int64  
 1   InvoiceID      230087 non-null  int64  
 2   StockItemID    230087 non-null  float64
 3   Description    229876 non-null  object 
 4   PackageTypeID  230087 non-null  int64  
 5   Quantity       230087 non-null  int64  
 6   UnitPrice      230087 non-null  float64
 7   TaxRate        230087 non-null  int64  
 8   TaxAmount      109408 non-null  float64
 9   LineProfit     230087 non-null  float64
 10  ExtendedPrice  112858 non-null  float64
 11  LastEditedBy   230087 non-null  int64  
 12  Customer       230087 non-null  object 
dtypes: float64(5), int64(6), object(2)
memory usage: 24.6+ MB


### I observed that the column TaxAmount also has many null values. This column indicates the amount of tax paid for each invoice line. The TaxAmount can be calculated using the following formula:

#TaxAmount = UnitPrice * Quantity * TaxRate / 100

### Based on this knowledge, my ecommendation would be to impute the missing values

In [ ]:
df['TaxAmount'] = df['TaxAmount'].fillna(df['UnitPrice'] * df['Quantity'] * df['TaxRate'] / 100)

### After imputing the data, there are no null values and the average TaxAmount hase gone from 113.29 to 112.95.

In [ ]:
df.describe()

InvoiceLineID      InvoiceID    StockItemID  PackageTypeID  \
count  230087.000000  230087.000000  230087.000000  230087.000000   
mean   114131.169266   35178.644621     110.181827       7.073468   
std     65894.369187   20337.134754      63.722514       0.644395   
min         1.000000       1.000000       1.000000       1.000000   
25%     57077.500000   17574.000000      54.000000       7.000000   
50%    114131.000000   35151.000000     111.000000       7.000000   
75%    171196.500000   52764.000000     165.000000       7.000000   
max    228265.000000   70510.000000     227.000000      10.000000   

            Quantity      UnitPrice        TaxRate      TaxAmount  \
count  230087.000000  230087.000000  230087.000000  230087.000000   
mean       39.247267      45.532654      14.977248     112.952167   
std        55.193231     139.461436       0.336518     219.486804   
min         1.000000       0.660000      10.000000       0.375000   
25%         5.000000      13.000000      15.000000      14.400000   
50%        10.000000      18.000000      15.000000      34.500000   
75%        60.000000      32.000000      15.000000     129.600000   
max      3150.000000    1899.000000      15.000000   14994.000000   

          LineProfit  ExtendedPrice   LastEditedBy  
count  230087.000000  112858.000000  230087.000000  
mean      375.314843     739.846178      10.799919  
std       753.072527    1704.618239       5.509919  
min      -645.000000    -999.000000       2.000000  
25%        51.000000      89.700000       6.000000  
50%       120.000000     248.400000      11.000000  
75%       390.000000     919.080000      16.000000  
max      9200.000000  114954.000000      20.000000

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 230087 entries, 0 to 230547
Data columns (total 13 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   InvoiceLineID  230087 non-null  int64  
 1   InvoiceID      230087 non-null  int64  
 2   StockItemID    230087 non-null  float64
 3   Description    229876 non-null  object 
 4   PackageTypeID  230087 non-null  int64  
 5   Quantity       230087 non-null  int64  
 6   UnitPrice      230087 non-null  float64
 7   TaxRate        230087 non-null  int64  
 8   TaxAmount      230087 non-null  float64
 9   LineProfit     230087 non-null  float64
 10  ExtendedPrice  112858 non-null  float64
 11  LastEditedBy   230087 non-null  int64  
 12  Customer       230087 non-null  object 
dtypes: float64(5), int64(6), object(2)
memory usage: 24.6+ MB


### The ExtendedPrice column, which indicates the total price paid by the customer, also has null values. The ExtendedPrice can be calculated using the following formula:

# ExtendedPrice = (UnitPrice * Quantity) + TaxAmount

### These null values can also be easily imputed.

In [ ]:
df['ExtendedPrice'] = df['ExtendedPrice'].fillna((df['UnitPrice'] * df['Quantity']) + df['TaxAmount'])

###We can see that rows with null values in the ExtendedPrice column have been updated.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 230087 entries, 0 to 230547
Data columns (total 13 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   InvoiceLineID  230087 non-null  int64  
 1   InvoiceID      230087 non-null  int64  
 2   StockItemID    230087 non-null  float64
 3   Description    229876 non-null  object 
 4   PackageTypeID  230087 non-null  int64  
 5   Quantity       230087 non-null  int64  
 6   UnitPrice      230087 non-null  float64
 7   TaxRate        230087 non-null  int64  
 8   TaxAmount      230087 non-null  float64
 9   LineProfit     230087 non-null  float64
 10  ExtendedPrice  230087 non-null  float64
 11  LastEditedBy   230087 non-null  int64  
 12  Customer       230087 non-null  object 
dtypes: float64(5), int64(6), object(2)
memory usage: 24.6+ MB


In [ ]:
df.describe()

InvoiceLineID      InvoiceID    StockItemID  PackageTypeID  \
count  230087.000000  230087.000000  230087.000000  230087.000000   
mean   114131.169266   35178.644621     110.181827       7.073468   
std     65894.369187   20337.134754      63.722514       0.644395   
min         1.000000       1.000000       1.000000       1.000000   
25%     57077.500000   17574.000000      54.000000       7.000000   
50%    114131.000000   35151.000000     111.000000       7.000000   
75%    171196.500000   52764.000000     165.000000       7.000000   
max    228265.000000   70510.000000     227.000000      10.000000   

            Quantity      UnitPrice        TaxRate      TaxAmount  \
count  230087.000000  230087.000000  230087.000000  230087.000000   
mean       39.247267      45.532654      14.977248     112.952167   
std        55.193231     139.461436       0.336518     219.486804   
min         1.000000       0.660000      10.000000       0.375000   
25%         5.000000      13.000000      15.000000      14.400000   
50%        10.000000      18.000000      15.000000      34.500000   
75%        60.000000      32.000000      15.000000     129.600000   
max      3150.000000    1899.000000      15.000000   14994.000000   

          LineProfit  ExtendedPrice   LastEditedBy  
count  230087.000000  230087.000000  230087.000000  
mean      375.314843     803.721183      10.799919  
std       753.072527    1688.000979       5.509919  
min      -645.000000    -999.000000       2.000000  
25%        51.000000     104.650000       6.000000  
50%       120.000000     257.600000      11.000000  
75%       390.000000     993.600000      16.000000  
max      9200.000000  114954.000000      20.000000

### I also noticed that some of the values in the ExtendedPrice column were values of -999. I applied the fix to those rows also.

In [ ]:
df.loc[df['ExtendedPrice'] == -999, 'ExtendedPrice'] = (df['UnitPrice'] * df['Quantity']) + df['TaxAmount']

In [ ]:
df.describe()

InvoiceLineID      InvoiceID    StockItemID  PackageTypeID  \
count  230087.000000  230087.000000  230087.000000  230087.000000   
mean   114131.169266   35178.644621     110.181827       7.073468   
std     65894.369187   20337.134754      63.722514       0.644395   
min         1.000000       1.000000       1.000000       1.000000   
25%     57077.500000   17574.000000      54.000000       7.000000   
50%    114131.000000   35151.000000     111.000000       7.000000   
75%    171196.500000   52764.000000     165.000000       7.000000   
max    228265.000000   70510.000000     227.000000      10.000000   

            Quantity      UnitPrice        TaxRate      TaxAmount  \
count  230087.000000  230087.000000  230087.000000  230087.000000   
mean       39.247267      45.532654      14.977248     112.952167   
std        55.193231     139.461436       0.336518     219.486804   
min         1.000000       0.660000      10.000000       0.375000   
25%         5.000000      13.000000      15.000000      14.400000   
50%        10.000000      18.000000      15.000000      34.500000   
75%        60.000000      32.000000      15.000000     129.600000   
max      3150.000000    1899.000000      15.000000   14994.000000   

          LineProfit  ExtendedPrice   LastEditedBy  
count  230087.000000  230087.000000  230087.000000  
mean      375.314843     867.637314      10.799919  
std       753.072527    1683.168093       5.509919  
min      -645.000000       2.875000       2.000000  
25%        51.000000     110.400000       6.000000  
50%       120.000000     264.500000      11.000000  
75%       390.000000     993.600000      16.000000  
max      9200.000000  114954.000000      20.000000

### The updated ExtendedPrice average is 867.64, up from 803.72.

### I will use the z-score method to determine how many outliers there are in the Quantity and ExtendedPrice columns. A thereshold of 3 will be used to determine which rows are outliers in these columns. I will get rows that have a z-score greater than or equal to 3.

In [ ]:
average_quantity = df['Quantity'].mean()
std_dev_quantity = df['Quantity'].std()

In [ ]:
df['Z-Score_quantity'] = (df['Quantity'] - average_quantity) / std_dev_quantity
df.head()

InvoiceLineID  InvoiceID  StockItemID  \
0              1          1         67.0   
1              2          2         50.0   
2              3          2         10.0   
3              4          3        114.0   
4              5          4        206.0   

                                         Description  PackageTypeID  Quantity  \
0           Ride on toy sedan car (Black) 1/12 scale              7        10   
1  Developer joke mug - old C developers never di...              7         9   
2               USB food flash drive - chocolate bar              7         9   
3                 Superhero action jacket (Blue) XXL              7         3   
4         Permanent marker black 5mm nib (Black) 5mm              7        96   

   UnitPrice  TaxRate  TaxAmount  LineProfit  ExtendedPrice  LastEditedBy  \
0      230.0       15     345.00       850.0        2645.00             7   
1       13.0       15      17.55        76.5         134.55             7   
2       32.0       15      43.20       180.0         331.20             7   
3       30.0       15      13.50        24.0         103.50             7   
4        2.7       15      38.88        96.0         298.08             7   

           Customer  Z-Score_quantity  
0     Stuff by Stew         -0.529907  
1   Wholesaler Plus         -0.548025  
2   Wholesaler Plus         -0.548025  
3   Big Buys Retail         -0.656734  
4  Terry's Trinkets          1.028255

In [ ]:
df.loc[df['Z-Score_quantity'] >= 3]

InvoiceLineID  InvoiceID  StockItemID  \
122               123         59        191.0   
197               198         95        191.0   
298               299        129        196.0   
305               306        132        186.0   
307               308        134        189.0   
...               ...        ...          ...   
230285         171205      52767        183.0   
230377          74832      23009        182.0   
230458         151681      46729        195.0   
230475         217008      67029        201.0   
230510         154951      47750        201.0   

                                              Description  PackageTypeID  \
122       Black and orange fragile despatch tape 48mmx75m              7   
197       Black and orange fragile despatch tape 48mmx75m              7   
298     Black and orange handle with care despatch tap...              7   
305                 Shipping carton (Brown) 457x457x457mm              7   
307                         Clear packaging tape 48mmx75m              7   
...                                                   ...            ...   
230285              Shipping carton (Brown) 480x270x320mm              7   
230377              Shipping carton (Brown) 457x279x279mm              7   
230458  Black and orange handle with care despatch tap...              7   
230475        Red and white urgent despatch tape 48mmx75m              7   
230510        Red and white urgent despatch tape 48mmx75m              7   

        Quantity  UnitPrice  TaxRate  TaxAmount  LineProfit  ExtendedPrice  \
122          288       3.70       15     159.84       547.2        1225.44   
197          252       3.70       15     139.86       478.8        1072.26   
298          216       4.10       15     132.84       453.6        1018.44   
305          250       2.10       15      78.75       250.0         603.75   
307          234       2.90       15     101.79       280.8         780.39   
...          ...        ...      ...        ...         ...            ...   
230285       250       2.74       15     102.75       385.0         787.75   
230377       250       1.28       15      48.00       170.0         368.00   
230458       216       3.70       15     119.88       518.4         919.08   
230475       240       3.70       15     133.20       504.0        1021.20   
230510       240       3.70       15     133.20       504.0        1021.20   

        LastEditedBy          Customer  Z-Score_quantity  
122                9  Terry's Trinkets          4.506943  
197                9     Stuff by Stew          3.854689  
298                2     Stuff by Stew          3.202435  
305                2   Big Buys Retail          3.818453  
307                2  Terry's Trinkets          3.528562  
...              ...               ...               ...  
230285            17  Terry's Trinkets          3.818453  
230377            10     Stuff by Stew          3.818453  
230458             5  Terry's Trinkets          3.202435  
230475            18     Stuff by Stew          3.637271  
230510            17   Big Buys Retail          3.637271  

[5987 rows x 14 columns]

In [ ]:
average_extendedprice = df['ExtendedPrice'].mean()
std_dev_extendedprice = df['ExtendedPrice'].std()

In [ ]:
df['Z-Score_extendedprice'] = (df['ExtendedPrice'] - average_extendedprice) / std_dev_extendedprice
df.head()

InvoiceLineID  InvoiceID  StockItemID  \
0              1          1         67.0   
1              2          2         50.0   
2              3          2         10.0   
3              4          3        114.0   
4              5          4        206.0   

                                         Description  PackageTypeID  Quantity  \
0           Ride on toy sedan car (Black) 1/12 scale              7        10   
1  Developer joke mug - old C developers never di...              7         9   
2               USB food flash drive - chocolate bar              7         9   
3                 Superhero action jacket (Blue) XXL              7         3   
4         Permanent marker black 5mm nib (Black) 5mm              7        96   

   UnitPrice  TaxRate  TaxAmount  LineProfit  ExtendedPrice  LastEditedBy  \
0      230.0       15     345.00       850.0        2645.00             7   
1       13.0       15      17.55        76.5         134.55             7   
2       32.0       15      43.20       180.0         331.20             7   
3       30.0       15      13.50        24.0         103.50             7   
4        2.7       15      38.88        96.0         298.08             7   

           Customer  Z-Score_quantity  Z-Score_extendedprice  
0     Stuff by Stew         -0.529907               1.055963  
1   Wholesaler Plus         -0.548025              -0.435540  
2   Wholesaler Plus         -0.548025              -0.318707  
3   Big Buys Retail         -0.656734              -0.453988  
4  Terry's Trinkets          1.028255              -0.338384

In [ ]:
df.loc[df['Z-Score_extendedprice'] >= 3]

InvoiceLineID  InvoiceID  StockItemID  \
96                 97         45        164.0   
111               112         53        170.0   
124               125         59        167.0   
256               257        116        167.0   
289               290        126        158.0   
...               ...        ...          ...   
230065         151133      46554        158.0   
230079         219963      67961        167.0   
230298         217447      67164        158.0   
230407         209037      64540        158.0   
230483           1708        598        215.0   

                                     Description  PackageTypeID  Quantity  \
96            32 mm Double sided bubble wrap 50m              7        50   
111     20 mm Anti static bubble wrap (Blue) 50m              7        90   
124     10 mm Anti static bubble wrap (Blue) 50m              7        60   
256     10 mm Anti static bubble wrap (Blue) 50m              7        70   
289           10 mm Double sided bubble wrap 50m              7        90   
...                                          ...            ...       ...   
230065        10 mm Double sided bubble wrap 50m              7        90   
230079  10 mm Anti static bubble wrap (Blue) 50m              7        80   
230298        10 mm Double sided bubble wrap 50m              7        60   
230407        10 mm Double sided bubble wrap 50m              7       100   
230483                Air cushion machine (Blue)              7         5   

        UnitPrice  TaxRate  TaxAmount  LineProfit  ExtendedPrice  \
96          112.0       15     840.00      2650.0        6440.00   
111         102.0       15    1377.00      4230.0       10557.00   
124          99.0       15     891.00      3240.0        6831.00   
256          99.0       15    1039.50      3780.0        7969.50   
289         105.0       15    1417.50      4410.0       10867.50   
...           ...      ...        ...         ...            ...   
230065      105.0       15    1417.50      4410.0       10867.50   
230079       99.0       15    1188.00      4320.0        9108.00   
230298      105.0       15     945.00      2940.0        7245.00   
230407      105.0       15    1575.00      4900.0       12075.00   
230483     1899.0       15    1424.25      3795.0       10919.25   

        LastEditedBy             Customer  Z-Score_quantity  \
96                 9        Stuff by Stew          0.194820   
111                9        Stuff by Stew          0.919546   
124                9     Terry's Trinkets          0.376001   
256                9        Stuff by Stew          0.557183   
289                9     Terry's Trinkets          0.919546   
...              ...                  ...               ...   
230065            13  Dennis Distribution          0.919546   
230079            19     Terry's Trinkets          0.738365   
230298             8     Terry's Trinkets          0.376001   
230407            17  Dennis Distribution          1.100728   
230483            15     Terry's Trinkets         -0.620498   

        Z-Score_extendedprice  
96                   3.310639  
111                  5.756622  
124                  3.542939  
256                  4.219343  
289                  5.941096  
...                       ...  
230065               5.941096  
230079               4.895746  
230298               3.788904  
230407               6.658493  
230483               5.971841  

[4522 rows x 15 columns]

### There are 5987 outliers in the Quantity column and 4522 in the ExtendedPrice column. I have made note of the outliers but do not recommend changing them or removing them from the dataset.

In [ ]:
df.head()

InvoiceLineID  InvoiceID  StockItemID  \
0              1          1         67.0   
1              2          2         50.0   
2              3          2         10.0   
3              4          3        114.0   
4              5          4        206.0   

                                         Description  PackageTypeID  Quantity  \
0           Ride on toy sedan car (Black) 1/12 scale              7        10   
1  Developer joke mug - old C developers never di...              7         9   
2               USB food flash drive - chocolate bar              7         9   
3                 Superhero action jacket (Blue) XXL              7         3   
4         Permanent marker black 5mm nib (Black) 5mm              7        96   

   UnitPrice  TaxRate  TaxAmount  LineProfit  ExtendedPrice  LastEditedBy  \
0      230.0       15     345.00       850.0        2645.00             7   
1       13.0       15      17.55        76.5         134.55             7   
2       32.0       15      43.20       180.0         331.20             7   
3       30.0       15      13.50        24.0         103.50             7   
4        2.7       15      38.88        96.0         298.08             7   

           Customer  Z-Score_quantity  Z-Score_extendedprice  
0     Stuff by Stew         -0.529907               1.055963  
1   Wholesaler Plus         -0.548025              -0.435540  
2   Wholesaler Plus         -0.548025              -0.318707  
3   Big Buys Retail         -0.656734              -0.453988  
4  Terry's Trinkets          1.028255              -0.338384

###The data set provided should have a unique ID for each invoice line, but there can be repeated invoice IDs. This is because several unique invoice lines can share the same invoice ID. Given this information, I will check for duplicates using only the InvoiceLineID column.

In [ ]:
df.loc[ df.duplicated(subset=['InvoiceLineID'])]

InvoiceLineID  InvoiceID  StockItemID  \
228265         199860      61652        222.0   
228266         197512      60915        161.0   
228267         113602      34985        167.0   
228268           6306       1950        184.0   
228269         134380      41395        155.0   
...               ...        ...          ...   
230543          36558      11285        139.0   
230544          21110       6502          9.0   
230545          34087      10504        179.0   
230546         103435      31836         20.0   
230547         192963      59503         70.0   

                                         Description  PackageTypeID  Quantity  \
228265                        Chocolate beetles 250g              1       216   
228266            20 mm Double sided bubble wrap 50m              7        50   
228267      10 mm Anti static bubble wrap (Blue) 50m              7        40   
228268         Shipping carton (Brown) 305x305x305mm              7       100   
228269               Large sized bubblewrap roll 50m              7        50   
...                                              ...            ...       ...   
230543                   Furry animal socks (Pink) M             10        36   
230544                 USB food flash drive - banana              7         2   
230545         Shipping carton (Brown) 229x229x229mm              7       125   
230546  DBA joke mug - you might be a DBA if (White)              7         1   
230547      Ride on toy sedan car (Green) 1/12 scale              7         5   

        UnitPrice  TaxRate  TaxAmount  LineProfit  ExtendedPrice  \
228265       8.55       10   184.6800       820.8      2031.4800   
228266     108.00       15   810.0000      4600.0      6210.0000   
228267      99.00       15   594.0000      2160.0      4554.0000   
228268       3.50       15    52.5000       160.0       402.5000   
228269      24.00       15   180.0000       550.0      1380.0000   
...           ...      ...        ...         ...            ...   
230543       5.00       15    27.0000       126.0       207.0000   
230544      32.00       15     9.6000        40.0        73.6000   
230545       1.05       15    19.6875        62.5       150.9375   
230546      13.00       15     1.9500         8.5        14.9500   
230547     230.00       15   172.5000       425.0      1322.5000   

        LastEditedBy          Customer  Z-Score_quantity  \
228265             5     Stuff by Stew          3.202435   
228266            12     Stuff by Stew          0.194820   
228267             4     Stuff by Stew          0.013638   
228268            16     Stuff by Stew          1.100728   
228269            20     Stuff by Stew          0.194820   
...              ...               ...               ...   
230543            20   Wholesaler Plus         -0.058835   
230544            10     Stuff by Stew         -0.674852   
230545            16  Terry's Trinkets          1.553682   
230546            20   Big Buys Retail         -0.692970   
230547             4     Stuff by Stew         -0.620498   

        Z-Score_extendedprice  
228265               0.691460  
228266               3.173992  
228267               2.190133  
228268              -0.276346  
228269               0.304404  
...                       ...  
230543              -0.392496  
230544              -0.471752  
230545              -0.425804  
230546              -0.506597  
230547               0.270242  

[2276 rows x 15 columns]

In [ ]:
df.drop_duplicates(subset=['InvoiceLineID'], inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 227811 entries, 0 to 229321
Data columns (total 15 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   InvoiceLineID          227811 non-null  int64  
 1   InvoiceID              227811 non-null  int64  
 2   StockItemID            227811 non-null  float64
 3   Description            227604 non-null  object 
 4   PackageTypeID          227811 non-null  int64  
 5   Quantity               227811 non-null  int64  
 6   UnitPrice              227811 non-null  float64
 7   TaxRate                227811 non-null  int64  
 8   TaxAmount              227811 non-null  float64
 9   LineProfit             227811 non-null  float64
 10  ExtendedPrice          227811 non-null  float64
 11  LastEditedBy           227811 non-null  int64  
 12  Customer               227811 non-null  object 
 13  Z-Score_quantity       227811 non-null  float64
 14  Z-Score_extendedprice  227811 non-null  f

### 2276 duplicates were found and dropped from the dataset leaving 227811 total rows.

### At this point, we have a data set that is mostly free of null values (only has some null values in the Description column). Those rows could be dropped prior to the data being run through a machine learning model but I will use this data set as is to answer the following questions.

### Which item has the highest total quantity sold?

#### Black and orange fragile despatch tape 48mmx75m.

In [ ]:
df.groupby('Description')['Quantity'].sum().sort_values(ascending=False)

Description
Black and orange fragile despatch tape 48mmx75m     207108
Black and orange fragile despatch tape 48mmx100m    193572
Clear packaging tape 48mmx75m                       158392
3 kg Courier post bag (White) 300x190x95mm          154525
Shipping carton (Brown) 356x356x279mm               151525
                                                     ...  
Superhero action jacket (Blue) L                      5450
Superhero action jacket (Blue) S                      5405
Superhero action jacket (Blue) XXL                    5404
DBA joke mug - mind if I join you? (Black)            5388
Superhero action jacket (Blue) XL                     5358
Name: Quantity, Length: 227, dtype: int64

### Which item has generated the most total profit?

####20 mm Double sided bubble wrap 50m.

In [ ]:
df.groupby('Description')['LineProfit'].sum().sort_values(ascending=False)

Description
20 mm Double sided bubble wrap 50m                    5283560.0
Air cushion machine (Blue)                            4418898.0
32 mm Anti static bubble wrap (Blue) 50m              3514220.0
10 mm Anti static bubble wrap (Blue) 50m              3439800.0
32 mm Double sided bubble wrap 50m                    2915530.0
                                                        ...    
Packing knife with metal insert blade (Yellow) 9mm      29888.1
Halloween zombie mask (Light Brown) L                  -63816.0
Halloween zombie mask (Light Brown) M                  -66444.0
Halloween zombie mask (Light Brown) S                  -70788.0
Halloween zombie mask (Light Brown) XL                 -72264.0
Name: LineProfit, Length: 227, dtype: float64

### Which item is purchased in greatest quantities in a single order, on average?

####Black and orange fragile despatch tape 48mmx75m.

In [ ]:
df.groupby('Description')['Quantity'].mean().sort_values(ascending=False)

Description
Black and orange fragile despatch tape 48mmx75m                                          199.526012
Black and orange fragile despatch tape 48mmx100m                                         199.148148
Clear packaging tape 48mmx75m                                                            145.313761
3 kg Courier post bag (White) 300x190x95mm                                               144.012116
Shipping carton (Brown) 356x356x279mm                                                    141.612150
                                                                                            ...    
USB food flash drive - hamburger                                                           5.327273
Superhero action jacket (Blue) L                                                           5.317073
DBA joke mug - SELECT caffeine FROM mug (Black)                                            5.283915
Developer joke mug - understanding recursion requires understanding recursion (Black)      5.277831
Superhero action jacket (Blue) XXL                                                         5.211186
Name: Quantity, Length: 227, dtype: float64

## Conclusion

Having clean data is very important whan perfoming an analysis.

I was able to impute much of the missing data and retain most of the rows.

Clean data makes the analysis easier and more reliable.